userspace/libscap/scap_savefile.c

```diff
@@ -2446,81 +2670,169 @@ int32_t scap_next_offline(scap_t *handle, OUT scap_evt **pevent, OUT uint16_t *p
-	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "unexpected block type %u", (uint32_t)bh.block_type);
-		handle->m_unexpected_block_readsize = readsize;
-		return SCAP_UNEXPECTED_BLOCK;
-	}
 
-	if(bh.block_total_length < sizeof(bh) + sizeof(struct ppm_evt_hdr) + 4)
-	{
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "block length too short %u", (uint32_t)bh.block_total_length);
-		return SCAP_FAILURE;
-	}
+		if((*pevent)->type >= PPM_EVENT_MAX)
+		{
+			//
+			// We're reading a capture that contains new syscalls.
+			// We can't do anything else that skips them.
+			//
+			continue;
+		}
 
-	//
-	// Read the event
-	//
-	readlen = bh.block_total_length - sizeof(bh);
-	if (readlen > FILE_READ_BUF_SIZE) {
-		snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "event block length %u greater than read buffer size %u",
-			 readlen,
-			 FILE_READ_BUF_SIZE);
-		return SCAP_FAILURE;
-	}
+		if(bh.block_type != EV_BLOCK_TYPE_V2 && bh.block_type != EVF_BLOCK_TYPE_V2)
+		{
+			//
+			// We're reading a old capture which events don't have nparams in the header.
+			// Convert it to the current version.
+			//
+			if((readlen + sizeof(uint32_t)) > FILE_READ_BUF_SIZE)
+			{
+				snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "cannot convert v1 event block to v2 (%lu greater than read buffer size %u)",
+					 readlen + sizeof(uint32_t),
+					 FILE_READ_BUF_SIZE);
+				return SCAP_FAILURE;
+			}
+			size_t offset =
+#ifdef PPM_ENABLE_SENTINEL
+				sizeof(uint32_t) + // sentinel
+#endif
+				sizeof(uint64_t) + // ts
+				sizeof(uint64_t) + // tid
+				sizeof(uint32_t) + // len
+				sizeof(uint16_t);  // type
 
-	readsize = gzread(f, handle->m_file_evt_buf, readlen);
-	CHECK_READ_SIZE(readsize, readlen);
+			memmove((char *)*pevent + offset + sizeof(uint32_t),
+				(char *)*pevent + offset,
+				readlen - ((char *)*pevent - handle->m_file_evt_buf) - offset);
+			(*pevent)->len += sizeof(uint32_t);
 
-	//
-	// EVF_BLOCK_TYPE has 32 bits of flags
-	//
-	*pcpuid = *(uint16_t *)handle->m_file_evt_buf;
+			//
+			// The number of parameters needs to be calculated based on the block len.
+			// Use the current number of parameters as starting point and decrease it
+			// until size matches.
+			//
+			char *end = (char *)*pevent + (*pevent)->len;
+			uint16_t *lens = (uint16_t *)((char *)*pevent + sizeof(struct ppm_evt_hdr));
+			uint32_t nparams;
+			for(nparams = g_event_info[(*pevent)->type].nparams; nparams >= 0; nparams--)
+			{
+				char *valptr = (char *)lens + nparams * sizeof(uint16_t);
+				if(valptr > end)
+				{
+					continue;
+				}
+				uint32_t i;
+				for(i = 0; i < nparams; i++)
+				{
+					valptr += lens[i];
+				}
+				if(valptr < end)
+				{
+					snprintf(handle->m_lasterr, SCAP_LASTERR_SIZE, "cannot convert v1 event block to v2 (corrupted trace file - can't calculate nparams).");
+					return SCAP_FAILURE;
+				}
+				ASSERT(valptr >= end);
+				if(valptr == end)
+				{
+					break;
+				}
+			}
+			(*pevent)->nparams = nparams;

```

Comment:
There's still something that sounds odd in this logic.

If I have an old event with 0 params while the current table has N, this should definitely be caught by keeping hitting `continue` in the loop above until we exit with `0`, and I'm ok with it.

However, what if the trace file is somehow corrupted and we end up returning `nparams` as `0` not because of the `continue`, but because every single combination we try is `valptr > end`? Can this happen in a legitimate case? If it cannot happen, we should handle the situation and throw an error, rather than putting it in the same bucket as the `continue` case, since in both cases we just end up exiting the loop with `nparams` set to `0`.

Reply:
(supposing that `valptr > end` was meant to be `valptr != end`)
I agree, it should never happen so I'll track down if `nparams` can't be calculated and return an error.